packages/server/src/plugin/subscriptionCallback/index.ts

```diff
@@ -0,0 +1,672 @@
+    : undefined;
+
+  return {
+    async requestDidStart({ request }) {
+      const subscriptionExtension = request?.extensions?.subscription;
+      // If it's not a callback subscription, ignore the request.
+      if (!subscriptionExtension) return;
+      const {
+        callback_url: callbackUrl,
+        subscription_id: id,
+        verifier,
+      } = subscriptionExtension;
+
+      return {
+        // Implementing `responseForOperation` is the only hook that allows us
+        // to bypass normal execution by returning our own response. We don't
+        // want Apollo Server to actually handle this subscription request, we
+        // want to handle everything ourselves. The actual subscription handling
+        // will be done in `willSendResponse`. The router expects the initial
+        // response to be a 200, with the `subscription-protocol` header set to
+        // `callback`.
+        async responseForOperation() {
+          logger?.debug('Received new subscription request', id);
+
+          return {
+            http: {
+              status: 200,
+              headers: new HeaderMap([['subscription-protocol', 'callback']]),
+            },
+            body: {
+              kind: 'single',
+              singleResult: {
+                data: null,
+              },
+            },
+          };
+        },
+        // The majority of the subscription work is implemented in this hook.
+        // This can _almost_ all be implemented in `responseForOperation`,
+        // however in the case of GraphQL validation errors,
+        // `responseForOperation` won't actually be called, but
+        // `willSendResponse` will.
+        async willSendResponse({
+          request,
+          schema,
+          document,
+          contextValue,
+          operationName,
+          response,
+        }) {
+          try {
+            logger?.debug('Sending `check` request to router', id);
+            // Before responding to the original request, we need to complete a
+            // roundtrip `check` request to the router, so we `await` this
+            // request.
+            await subscriptionManager.checkRequest({
+              callbackUrl,
+              id,
+              verifier,
+            });
+            logger?.debug('`check` request successful', id);
+          } catch (e) {
+            const graphqlError = ensureGraphQLError(e);
+            logger?.error(
+              `\`check\` request failed: ${graphqlError.message}`,
+              id,
+            );
+            // In the event of a check failure, we respond to the original
+            // request with a >=400 status code.
+            if (response.body.kind === 'single') {
+              response.body.singleResult.errors = [graphqlError];
+              response.http.status = 500;
+            }
+            return;
+          }
+
+          // The `check` request was successful, so we can initialize the actual
```

Comment:
I wouldn't expect the router to send duplicates but I suppose it could, are you suggesting defending against that (maybe by checking to see if we're already subscribed on that id in the `SubscriptionManager`)?

Reply:
I guess I was thinking about a client that was generating non-unique subscription ids.